fs/util.cc

```diff
@@ -16,143 +13,151 @@
 
 namespace ROCKSDB_NAMESPACE {
 
-std::unique_ptr<ZonedBlockDevice> zbd_open(std::string const &zbd_path,
-                                           bool readonly, bool exclusive) {
-  std::unique_ptr<ZonedBlockDevice> zbd{
-      new ZonedBlockDevice(zbd_path, nullptr)};
-  IOStatus open_status = zbd->Open(readonly, exclusive);
+IOStatus zenfs_zbd_open(std::filesystem::path const &path, bool readonly,
+                        bool exclusive,
+                        std::unique_ptr<ZonedBlockDevice> &out_zbd) {
+  out_zbd =
+      std::unique_ptr<ZonedBlockDevice>{new ZonedBlockDevice(path, nullptr)};
+  IOStatus status = out_zbd->Open(readonly, exclusive);
 
-  if (!open_status.ok()) {
+  if (!status.ok()) {
     fprintf(stderr, "Failed to open zoned block device: %s, error: %s\n",
-            zbd_path.c_str(), open_status.ToString().c_str());
-    zbd.reset();
+            path.c_str(), status.ToString().c_str());
+    out_zbd.reset();
```

Comment:
Ok, I see that.
But we reset it only when status is not ok.
And in calling context, if status is not ok, we anyway return which would destruct the zbd object, right ?

Reply:
We reset the smart pointer on the error path in order to not return a partially initialized object to the caller. You are right, the caller should never access the object if the returned status is not OK. The object would be destroyed when the unique_ptr goes out of scope in the calling context, but only if zbd is a reference to a variable in the calling context. It might be from higher in the call stack or even on the heap, we cannot know. By calling reset, we ensure object destruction before function return, and we ensure that the unique_ptr will be empty, wherever it is allocated. 

We could state in the API contract that the value of the zbd argument is undefined and should not be used in case status is not OK. But then we would rely on the caller doing something that we could just assert before returning. Resetting it before return on the error path is a defensive programming strategy that makes it more probable that the API is used the way we intend.

Also, in case the caller wants to retry the call, it should not be the task of the caller to destroy a partially initialized object from the zbd argument. Better to keep the invariant that if an object is returned, it is fully initialized.

Reply:
If `zbd` was a regular pointer pointer, we would also set that to `nullptr`.